### ISSUE TITLE
BRAT Parameter Errors

### ISSUE THREAD

**Issue by philipbaileynar** [Riverscapes/riverscapes-tools#445]
These are the errors I now get pull your latest BRAT parameter changes. It should be fairly self explanatory... but here's a summary in plain English:

- HUCs 10050004, 10100005 and 17010103 are missing a watershed hydro parameter value for parameter ACWIDTH.
- HUC 10050004 is missing a watershed hydro param for  ELEV.

----

## Hydrologic equation Error
```
Equation: 0.026*(ACWIDTH**1.667) 
Exception: TypeError("'NoneType' object is not subscriptable")
Watersheds:
        10050004:
                FOREST: 0.4312265
                PRECIP: 392.12903
                FORESTP1: 1.4312265
                DRNAREA: 1.0
        10100005:
                FOREST: 0.3431033
                PRECIP: 437.20422
                FORESTP1: 1.3431033
                DRNAREA: 1.0
```

## Hydrologic equation Error
```
Equation: 0.0521*(ACWIDTH**1.58) 
Exception: TypeError("'NoneType' object is not subscriptable")
Watersheds:
        17010103:
                ELEV: 1373.4336
                FOREST: 36.28387
                RR: 1792.186
                PRECIP: 1011.9015
                SLOP30_30M: 2.099814
                MINELEV: 557.51636
                MEANSLOPE: 13.645659
                CANOPY_PCT: 25.27202
                ELEV1K: 1373.4336
                FORESTP1: 37.28387
                SOILCAPACITY: 75.11885
                JANMAXT2K: 28.653128
                JANMINT2K: 15.658107
                DRNAREA: 1.0
```

## Hydrologic equation Error
```
Equation: 30.5*(DRNAREA**0.601)* ((ELEV/1000)**-0.913) 
Exception: TypeError("'NoneType' object is not subscriptable")
Watersheds:
        10050004:
                FOREST: 0.4312265
                PRECIP: 392.12903
                FORESTP1: 1.4312265
                DRNAREA: 1.0
```
Errors occurred: Aborting due to 3 hydrology equation errors

**Comment by Cashe93** [comment]
@philipbaileynar, after speaking with @leallysmith I think we are on the right track but have a couple of questions before we can continue. ACWIDTH is a new parameter and we are unsure how to proceed; is this something that can be automated on a reach-by-reach basis, or do we just need to figure out a way to calculate the average width per HUC?

**Comment by philipbaileynar** [comment]
First, it's OK if you need new parameters that aren't already defined. The new parameter must be added to the `hydro_params` table before you attempt to store any values in the `watershed_hydro_params` table. So if you truly need `ACWIDTH` then add it to `hydro_params` and note down the new `param_id` for use when you add watershed values.

Second, I know Tyler would disregard certain equations because they had parameters for which it would be hard to calculate values. I know that sounds like cheating, but a) we only need rough discharge and b) this is all pretty crude science regardless. So you might borrow equations from neighbouring HUCs when you get stuck.

Third, you can try using [StreamStats](https://streamstats.usgs.gov/ss/) to calculate discharge. I recall there's a checkbox for having the output show equations. You might find an alternative equation for nearby streams. 

Fourth, this whole approach requires the equations to be "self-contained". This process requires all the values (except `DRNAREA`) to be stored in the database. i.e. we don't have flexibility to tap into other GIS processes or calculations for parameter values.

Does that help?

**Comment by Cashe93** [comment]
Thanks, @philipbaileynar! That does help. Slowly but surely making progress. I'll try to have a handful of test runs for you by midday tomorrow.

**Comment by Cashe93** [comment]
@philipbaileynar, within the hydro params, is the column named 'data_units' the units that BRAT uses? So, if we had an equation with a parameter that uses feet per mile, would the 'data_units' be meters per kilometer? And the conversion in this case be 0.189394?

**Comment by philipbaileynar** [comment]
- `data_units` refers to the units of the values in the `watershed_hydro_params` table.
- `equation_units` are the units of the variable (parameter) in the hydrological equation.
- `conversion` is the multiplier that is applied to the value in `watershed_hydro_params` table prior to applying the parameter value to the equation. This allows you to store a parameter value that has different units than the equation (e.g. store elevation values in feet, but the equation uses meters).

Make sense?

**Comment by Cashe93** [comment]
That does make sense, but for clarity: 

I have a parameter that multiplies relief by stream density and the equation expects this to be feet per mile. So if I'm using a 10 meter DEM to derive the relief value, would I need to convert feet per mile to meters per km?

**Comment by philipbaileynar** [comment]
You can store the parameter value as feet per km and use a conversion factor to ensure that the parameter value is converted for use in the equation.

**Comment by philipbaileynar** [comment]
@Cashe93 I don't think you need to manually calculate any parameters. You should be able to get them from streamstats. 

Use streamstats to calculate discharge for any stream in a HUC and the output should show you the parameter values for that hydrological region. 

It should be quick per HUC. Don't spend time calculating params. 

Let me know if that doesn't make sense.

**Comment by Cashe93** [comment]
@philipbaileynar, I've never used streamstats so I'm not sure how that is done. I've been able to locate documentation through stream stats, but not specific parameter values.

**Comment by Cashe93** [comment]
Scratch that, I figured it out. However, for the Dakotas there are no low flow equations listed, so you can't use streamstats to find those parameters. I'm not sure of another way to do this other than manually. Please advise if there is a way.

**Comment by philipbaileynar** [comment]
I have seen this "missing low flow equation" problem before. My advice would be to find another "arid" area and copy the low flow equation for this HUC and use it anywhere there are no low flow equation. Then put something meaningful in the notes fields for the watershed so that we can go back and identify these places later.

Maybe look at HUCs in the NE of Idaho and see if there is a common equation used for several HUCs?

**Comment by Cashe93** [comment]
I found one from NE Minnesota. We'll see how that works. There should now be three North Dakota HUCS ready for testing:

- 10060007
- 09020315
- 09020201

If these work, the rest of North Dakota should go pretty smoothly. I'll be out of town next week but will pick back up on this when I get back into the lab.

**Comment by philipbaileynar** [comment]
@Cashe93 I think you are missing a closing brace on the end of the equation for 10060007:



## Hydrologic equation Error
```
Equation: (10**−0.137)*(DRNAREA**0.510)*(RUGGED**0.630)*((10**(−0.139*COMPRAT))
Exception: SyntaxError('invalid character in identifier', ('<string>', 1, 7, '(10**−0.137)*(DRNAREA**0.510)*(RUGGED**0.630)*((10**(−0.139*COMPRAT))'))
Watersheds:
        10060007:
                FOREST: 0.0
                PRECIP: 431.64554
                MEANSLOPE: 3.401036
                FORESTP1: 1.0
                SOM: 5.0
                RUGGED: 0.0
                COMPRAT: 2.95
                DRNAREA: 1.0
```

**Comment by philipbaileynar** [comment]
@leallysmith I am seeing a lot of hydrology equation errors. Many involve a mistmatch in opening and closing braces. Like the following 7 open braces and 6 closing braces:

```python
((1.897*(10**−8))*(DRNAREA**1.684)*((SOM+1)**2.546)*(MEANSLOPE**3.664)
```

This occurs in the following HUCs:

| watershed\_id |
| :--- |
| 09020307 |
| 09020308 |
| 09020109 |
| 09020105 |
| 09020316 |
| 09020310 |
| 09020315 |
| 09020104 |
| 09020301 |
| 09020303 |
| 09020311 |
| 09020306 |

**Comment by wally-mac** [comment]
@philipbaileynar can this now be closed?
